sklearn/manifold/t_sne.py

```diff
@@ -477,29 +773,53 @@ def _tsne(self, P, alpha, n_samples, random_state, X_embedded=None):
                                                    self.n_components)
         params = X_embedded.ravel()
 
+        opt_args = {}
+        opt_args = {"n_iter": 50, "momentum": 0.5, "it": 0,
+                    "learning_rate": self.learning_rate,
+                    "verbose": self.verbose, "n_iter_check": 25,
+                    "kwargs": dict(skip_num_points=skip_num_points)}
+        if self.method == 'barnes_hut':
+            m = "Must provide an array of neighbors to use Barnes-Hut"
+            assert neighbors is not None, m
+            obj_func = _kl_divergence_bh
+            objective_error = _kl_divergence_error
+            sP = squareform(P).astype(np.float32)
+            neighbors = neighbors.astype(np.int64)
+            args = [sP, neighbors, degrees_of_freedom, n_samples,
+                    self.n_components]
+            opt_args['args'] = args
+            opt_args['min_grad_norm'] = 1e-3
+            opt_args['n_iter_without_progress'] = 30
+            # Don't always calculate the cost since that calculation
+            # can be nearly as expensive as the gradient
+            opt_args['objective_error'] = objective_error
+            opt_args['kwargs']['angle'] = self.angle
+            opt_args['kwargs']['verbose'] = self.verbose
+        else:
+            obj_func = _kl_divergence
+            opt_args['args'] = [P, degrees_of_freedom, n_samples,
+                                self.n_components]
+            opt_args['min_error_diff'] = 0.0
+            opt_args['min_grad_norm'] = 0.0
+
         # Early exaggeration
         P *= self.early_exaggeration
-        params, error, it = _gradient_descent(
-            _kl_divergence, params, it=0, n_iter=50, momentum=0.5,
-            min_grad_norm=0.0, min_error_diff=0.0,
-            learning_rate=self.learning_rate, verbose=self.verbose,
-            args=[P, alpha, n_samples, self.n_components])
-        params, error, it = _gradient_descent(
-            _kl_divergence, params, it=it + 1, n_iter=100, momentum=0.8,
-            min_grad_norm=0.0, min_error_diff=0.0,
-            learning_rate=self.learning_rate, verbose=self.verbose,
-            args=[P, alpha, n_samples, self.n_components])
+        params, error, it = _gradient_descent(obj_func, params, **opt_args)
+        opt_args['n_iter'] = 100
+        opt_args['momentum'] = 0.8
+        opt_args['it'] = it + 1
+        params, error, it = _gradient_descent(obj_func, params, **opt_args)
         if self.verbose:
             print("[t-SNE] Error after %d iterations with early "
                   "exaggeration: %f" % (it + 1, error))
+        # Save the final number of iterations
+        self.n_iter_final = it
```

Comment:
I actually it should be named `n_iter_` to be consistent with the conventions of the project.